### ISSUE TITLE
Application abort when using amqp destination

### ISSUE THREAD

**Issue by freak82** [balabit/syslog-ng#1513]
## Version of syslog-ng

```
[/etc/syslog-ng]# syslog-ng --version
syslog-ng 3.9.1
Installer-Version: 3.9.1
Revision: 
Module-Directory: //usr/lib64/syslog-ng
Module-Path: //usr/lib64/syslog-ng
Available-Modules: add-contextual-data,afamqp,affile,afprog,afsocket,afstomp,afuser,basicfuncs,cef,confgen,cryptofuncs,csvparser,date,dbparser,disk-buffer,graphite,json-plugin,kvformat,linux-kmsg-format,pseudofile,sdjournal,syslogformat,system-source
Enable-Debug: off
Enable-GProf: off
Enable-Memtrace: off
Enable-IPv6: on
Enable-Spoof-Source: on
Enable-TCP-Wrapper: on
Enable-Linux-Caps: on

```

## Platform

```
[/etc/syslog-ng]# cat /etc/centos-release
CentOS Linux release 7.3.1611 (Core) 
[/etc/syslog-ng]# uname -a
Linux c7-syslog-mcn-1 3.10.0-514.21.1.el7.x86_64 #1 SMP Thu May 25 17:04:51 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

## Failure

The syslog-ng deamon aborts when a amqp destination, to Azure Event Hub, is added.

```
May 31 19:30:24 c7-syslog-mcn-1 systemd: Starting System Logger Daemon...
May 31 19:30:24 c7-syslog-mcn-1 systemd: Started System Logger Daemon.
May 31 19:30:41 c7-syslog-mcn-1 syslog-ng: iv_timer_register: called with timer still on the heap
May 31 19:30:41 c7-syslog-mcn-1 systemd: syslog-ng.service: main process exited, code=killed, status=6/ABRT
May 31 19:30:41 c7-syslog-mcn-1 systemd: Unit syslog-ng.service entered failed state.
May 31 19:30:41 c7-syslog-mcn-1 systemd: syslog-ng.service failed.
May 31 19:30:42 c7-syslog-mcn-1 systemd: syslog-ng.service holdoff time over, scheduling restart.
May 31 19:30:42 c7-syslog-mcn-1 systemd: Starting System Logger Daemon...
May 31 19:30:42 c7-syslog-mcn-1 systemd: Started System Logger Daemon.
May 31 19:30:51 c7-syslog-mcn-1 syslog-ng: iv_timer_register: called with timer still on the heap
May 31 19:30:51 c7-syslog-mcn-1 systemd: syslog-ng.service: main process exited, code=killed, status=6/ABRT
May 31 19:30:51 c7-syslog-mcn-1 systemd: Unit syslog-ng.service entered failed state.
May 31 19:30:51 c7-syslog-mcn-1 systemd: syslog-ng.service failed.
May 31 19:30:52 c7-syslog-mcn-1 systemd: syslog-ng.service holdoff time over, scheduling restart.
May 31 19:30:52 c7-syslog-mcn-1 systemd: Starting System Logger Daemon...
May 31 19:30:52 c7-syslog-mcn-1 systemd: Started System Logger Daemon.
May 31 19:31:01 c7-syslog-mcn-1 syslog-ng: iv_timer_register: called with timer still on the heap
May 31 19:31:01 c7-syslog-mcn-1 systemd: syslog-ng.service: main process exited, code=killed, status=6/ABRT
May 31 19:31:01 c7-syslog-mcn-1 systemd: Unit syslog-ng.service entered failed state.
May 31 19:31:01 c7-syslog-mcn-1 systemd: syslog-ng.service failed.
May 31 19:31:02 c7-syslog-mcn-1 systemd: syslog-ng.service holdoff time over, scheduling restart.
May 31 19:31:02 c7-syslog-mcn-1 systemd: Starting System Logger Daemon...
May 31 19:31:02 c7-syslog-mcn-1 systemd: Started System Logger Daemon.
^C

```

## Steps to reproduce

1. start syslog-ng

2. it aborts few seconds later

## Configuration

```
@version:3.9
@include "scl.conf"

options {
    flush_lines (0);
    time_reopen (10);
    log_fifo_size (1000);
    chain_hostnames (off);
    use_dns (no);
    use_fqdn (no);
    create_dirs (no);
    keep_hostname (yes);
};

# Sorry I can't provide some details as ips, credentials, etc.
source s_network {
    network(ip(xxx.xxx.xxx.xxx) port(514) transport("tcp"));
};

destination d_amqp {
    amqp(
        vhost("/")
        host("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx")
        port(5672)
        username("xxxxxxxxx")
        password("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx")
        exchange("syslog")
        exchange-type("fanout")
        #routing-key("")
        #body("")
        #persistent(yes)
    );
};


```

**Comment by LarryWax** [comment]
Start syslog-ng from command line with "syslog-ng -Fvdte" and copy output here

**Comment by bazsi** [comment]
Hmm, it seems a timer is registered before it has expired.

Can you give me the ivykis version as well? Where did you get the binaries?

On Jun 6, 2017 16:51, "LarryWax" <notifications@github.com> wrote:

Start syslog-ng from command line with "syslog-ng -Fvdte" and copy output
here

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/balabit/syslog-ng/issues/1513#issuecomment-306530429>,
or mute the thread

**Comment by freak82** [comment]
Sorry, I really don't have time to do the required syslog-ng test. I'll do it as soon as I can.

For the installation I followed the steps described here:
https://www.balabit.com/blog/installing-latest-syslog-ng-on-rhel-and-other-rpm-distributions/?_ga=2.24217658.1029445652.1496821787-151667521.1496215816

cd /etc/yum.repos.d/
wget https://copr.fedorainfracloud.org/coprs/czanik/syslog-ng39/repo/epel-7/czanik-syslog-ng39-epel-7.repo
yum install syslog-ng
systemctl enable syslog-ng
systemctl start syslog-ng

**Comment by mitzkia** [comment]
I think I have a reproduction for this issue. 
In my case syslog-ng aborted when the username and password was not correct for amqp destination.
1, start netcat (or real amqp destination) with: `nc -lk 127.0.0.1 5672`
2, start syslog-ng with the following config
```
@version: 3.9
options{ stats_level(3); time_reopen(1); };
source s_network { network( ip("127.0.0.1") port(10001)); };
destination d_amqp { 
  amqp(
    host(127.0.0.1)
    port(5672)
    username("a")
    password("a")
  );
};
log { source(s_network); destination(d_amqp); };
```
3, send 1 message to syslog-ng with: `bin/loggen -n 1 -i 127.0.0.1 10001`
4, syslog-ng console log
```
[2017-06-27T13:41:47.401449] Syslog connection accepted; fd='15', client='AF_INET(127.0.0.1:36594)', local='AF_INET(127.0.0.1:10001)'
[2017-06-27T13:41:47.401845] Incoming log entry; line='<38>2017-06-27T13:41:47 localhost prg00000[1234]: seq: 0000000000, thread: 0000, runid: 1498563707, stamp: 2017-06-27T13:41:47 PADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADDPADD'
[2017-06-27T13:41:47.402182] Setting value; msg='0x7f1240014600', name='HOST_FROM', value='localhost'
[2017-06-27T13:41:47.402218] Setting value; msg='0x7f1240014600', name='HOST', value='localhost'
[2017-06-27T13:41:47.402283] Setting value; msg='0x7f1240014600', name='SOURCE', value='s_network'
[2017-06-27T13:41:47.402334] EOF occurred while reading; fd='15'
[2017-06-27T13:41:47.402433] Syslog connection closed; fd='15', client='AF_INET(127.0.0.1:36594)', local='AF_INET(127.0.0.1:10001)'
[2017-06-27T13:41:47.402582] Closing log transport fd; fd='15'
[2017-06-27T13:41:48.303599] Error during AMQP login; driver='d_amqp#0', error='connection closed unexpectedly', time_reopen='1'
[2017-06-27T13:41:51.316612] Error during AMQP login; driver='d_amqp#0', error='connection closed unexpectedly', time_reopen='1'
iv_timer_register: called with timer still on the heap
Aborted (core dumped)
```

backtrace:
```
(gdb) bt full
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:58
        set = {__val = {0, 140077066767008, 140077066003024, 7465859390, 140077066766152, 140077066003008, 140077061667633, 252833149, 4294967295, 0, 140077052202456, 140077066770024, 0, 
            4294967656, 140076936020923, 9}}
        pid = <optimized out>
        tid = <optimized out>
#1  0x00007f663af3137a in __GI_abort () at abort.c:89
        save_stage = 2
        act = {__sigaction_handler = {sa_handler = 0x2c27302370716d61, sa_sigaction = 0x2c27302370716d61}, sa_mask = {__val = {2827541964045575456, 140076936020923, 140077056098560, 0, 0, 
              0, 3432, 140077066003552, 140077062150826, 55, 55, 140077056120656, 0, 140077065993360, 140077066016512, 55}}, sa_flags = 998383593, sa_restorer = 0x7f663bc2a460}
        sigs = {__val = {32, 0 <repeats 15 times>}}
#2  0x00007f663b879ec3 in iv_fatal (fmt=fmt@entry=0x7f663b89f9a0 "iv_timer_register: called with timer still on the heap") at ../../../../lib/ivykis/src/iv_fatal.c:57
        ap = {{gp_offset = 8, fp_offset = 48, overflow_arg_area = 0x7f663bc2a930, reg_save_area = 0x7f663bc2a870}}
        msg = "iv_timer_register: called with timer still on the heap\000\000\061\033\210;f\177\000\000\060\246\302;f\177\000\000\001\000\000\000\000\000\000\000l\000\000\000\000\000\000\000\206&\001;f\177\000\000\001\200\255\373\000\000\000\000\203\247\302;f\177\000\000\203\247\302;f\177\000\000\203\247\302;f\177\000\000\203\247\302;f\177\000\000\212\247\302;f\177\000\000\357\247\302;f\177\000\000\203\247\302;f\177\000\000\357\247\302;f\177", '\000' <repeats 42 times>...
#3  0x00007f663b87a405 in iv_timer_register (_t=0x55c674e237d8) at ../../../../lib/ivykis/src/iv_timer.c:239
        t = 0x55c674e237d8
        p = <optimized out>
        index = <optimized out>
#4  0x00007f663b8351aa in log_threaded_dest_driver_suspend () from /home/micek/micek_work/git_projects/github_balabit_syslog_ng_ose_master/install/lib/libsyslog-ng-3.10.so.0
No symbol table info available.
#5  0x00007f663b8353f3 in _disconnect_and_suspend () from /home/micek/micek_work/git_projects/github_balabit_syslog_ng_ose_master/install/lib/libsyslog-ng-3.10.so.0
No symbol table info available.
#6  0x00007f663b835611 in log_threaded_dest_driver_do_insert () from /home/micek/micek_work/git_projects/github_balabit_syslog_ng_ose_master/install/lib/libsyslog-ng-3.10.so.0
No symbol table info available.
#7  0x00007f663b835783 in log_threaded_dest_driver_do_work () from /home/micek/micek_work/git_projects/github_balabit_syslog_ng_ose_master/install/lib/libsyslog-ng-3.10.so.0
No symbol table info available.
#8  0x00007f663b879f93 in iv_run_tasks (st=st@entry=0x7f6634002610) at ../../../../lib/ivykis/src/iv_task.c:48
        t = <optimized out>
        tasks = {next = 0x7f663bc2aa30, prev = 0x7f663bc2aa30}
#9  0x00007f663b87c25c in iv_main () at ../../../../lib/ivykis/src/iv_main_posix.c:106
        to = {tv_sec = 3600, tv_nsec = 0}
#10 0x00007f663b835a16 in log_threaded_dest_driver_worker_thread_main () from /home/micek/micek_work/git_projects/github_balabit_syslog_ng_ose_master/install/lib/libsyslog-ng-3.10.so.0
No symbol table info available.
#11 0x00007f663b83b99b in _worker_thread_func () from /home/micek/micek_work/git_projects/github_balabit_syslog_ng_ose_master/install/lib/libsyslog-ng-3.10.so.0
No symbol table info available.
#12 0x00007f663b550f65 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
No symbol table info available.
#13 0x00007f663b2c86da in start_thread (arg=0x7f663bc2d700) at pthread_create.c:456
        __res = <optimized out>
        pd = 0x7f663bc2d700
        now = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140077066016512, 6219136527917614170, 140723687776078, 140723687776079, 140077066017216, 140077066016512, -6304273953775975334, 
                -6304272148972862374}, mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
        pagesize_m1 = <optimized out>
        sp = <optimized out>
        freesize = <optimized out>
        __PRETTY_FUNCTION__ = "start_thread"
#14 0x00007f663b002d7f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:105
No locals.
```

**Comment by mitzkia** [comment]
Ok, verified, #1573 fixes this issue. Can not reproduce the crash with the above steps.
